### ISSUE TITLE
sha256 regression in master 2019-03-24

### ISSUE THREAD

**Issue by petrovr** [openssl/openssl#8565]
last tested "good" version : master-20190317

PKIX-SSH regression test try-ciphers start to fail with current master branch. Quote from test output:
...
test try ciphers: cipher aes256-ctr mac hmac-sha1
test try ciphers: cipher aes256-ctr mac hmac-sha1-96
test try ciphers: cipher aes256-ctr mac hmac-sha2-256
_ssh failed with mac hmac-sha2-256 cipher aes256-ctr_
test try ciphers: cipher aes256-ctr mac hmac-sha2-512
test try ciphers: cipher aes256-ctr mac hmac-md5
test try ciphers: cipher aes256-ctr mac hmac-md5-96
test try ciphers: cipher aes256-ctr mac umac-64@openssh.com
test try ciphers: cipher aes256-ctr mac umac-128@openssh.com
test try ciphers: cipher aes256-ctr mac hmac-sha1-etm@openssh.com
test try ciphers: cipher aes256-ctr mac hmac-sha1-96-etm@openssh.com
test try ciphers: cipher aes256-ctr mac hmac-sha2-256-etm@openssh.com
_ssh failed with mac hmac-sha2-256-etm@openssh.com cipher aes256-ctr_
test try ciphers: cipher aes256-ctr mac hmac-sha2-512-etm@openssh.com
test try ciphers: cipher aes256-ctr mac hmac-md5-etm@openssh.com
test try ciphers: cipher aes256-ctr mac hmac-md5-96-etm@openssh.com
test try ciphers: cipher aes256-ctr mac umac-64-etm@openssh.com
test try ciphers: cipher aes256-ctr mac umac-128-etm@openssh.com
....

**Comment by mattcaswell** [comment]
Can you supply some instructions on the steps I need to take to reproduce the above against master OpenSSL? Most likely this is related to de29ff17a20.

**Comment by mattcaswell** [comment]
Having downloaded the PKIX-SSH code from gitlab, I compiled OpenSSL:
````
./config -d --strict-warnings --prefix=/usr/local/ssl
````
Then ran "make" and "make install"

In PKIX-SSH source dir I ran:

````
./configure --with-ssl-dir=/usr/local/ssl
````
But I get...
````
checking how to link programs using OpenSSL functions... -lcrypto 
checking for OPENSSL_init_crypto... yes
checking for OPENSSL_version_major... yes
checking OpenSSL header version... 3.0.0 (OpenSSL 3.0.0-dev xx XXX xxxx)
checking OpenSSL library version... not found
configure: error: OpenSSL library not found.
````
I guess I'm doing something stupid - I also don't know how to run the test above once I've successfully configured and run "make" (or does that happen as part of make?).

**Comment by levitte** [comment]
`--with-ssl-dir=/usr/local/ssl/lib`

**Comment by mattcaswell** [comment]
Nope. That fails at the first of the above checks with:

````
checking how to link programs using OpenSSL functions... fail
configure: error: cannot link with OpenSSL crypto library
`````

**Comment by levitte** [comment]
In that case, there's trouble with this test:

https://gitlab.com/secsh/pkixssh/blob/master/configure.ac#L2705

**Comment by petrovr** [comment]
Hi Matt,

On openssl issues I work only in my limited spare time. This mean that 
some time I could provide information after a number of days.

mattcaswell wrote:
> Having downloaded the PKIX-SSH code from gitlab, I compiled OpenSSL:
> ````
> ./config -d --strict-warnings --prefix=/usr/local/ssl
> ````
> Then ran "make" and "make install"
> 
> In PKIX-SSH source dir I ran:
> 
> ````
> ./configure --with-ssl-dir=/usr/local/ssl
> ````
> But I get...
> ````
> checking how to link programs using OpenSSL functions... -lcrypto
> checking for OPENSSL_init_crypto... yes
> checking for OPENSSL_version_major... yes
> checking OpenSSL header version... 3.0.0 (OpenSSL 3.0.0-dev xx XXX xxxx)
> checking OpenSSL library version... not found
> configure: error: OpenSSL library not found.
> ````
> I guess I'm doing something stupid - I also don't know how to run the test above once I've successfully configured and run "make" (or does that happen as part of make?).
> 


Building outside regular system paths requires some system tuning.
I don't know how is build openssl above mainly RUNPATH configuration.


Let see may configuration     ./config.status --config:

'--prefix=[INSTALLROOT]/usr' \
'--sysconfdir=[INSTALLROOT]/etc/ssh' \
'--with-pid-dir=[INSTALLROOT]/var/run' \
'--without-pam' \
'--with-tcp-wrappers' \
'--with-md5-passwords' \
'--with-default-path=[INSTALLROOT]/usr/sbin:[INSTALLROOT]/usr/bin:/usr/local/sbin:/usr/sbin:/sbin:/usr/local/bin:/usr/bin:/bin' 
'--enable-ldap' '--with-ldap-prefix=/opt/openldap/2.4.46-sslmaster/usr' \
'--with-ldap-libdir=/opt/openldap/2.4.46-sslmaster/usr/lib64' \
'--with-ldap-sysconfdir=/opt/openldap/2.4.46-sslmaster/etc/openldap' \
'OPENSSL=/usr/local/openssl64/master/bin/openssl' \
'--build=x86_64-slackware-linux' \
'build_alias=x86_64-slackware-linux' \
'CFLAGS= -g -O3 -std=c99 -fstack-protector-all -Wall -Wextra -Wpedantic 
-Wmultichar -Wundef -Wshadow -Wbad-function-cast -Wcast-align 
-Wno-sign-compare -Waggregate-return -Wstrict-prototypes 
-Wmissing-prototypes -Wmissing-declarations -Winline -Wno-long-long' \
'LDFLAGS= -L/usr/local/openssl64/master/lib 
-Wl,-rpath,/usr/local/openssl64/master/lib 
-Wl,-rpath,/opt/openldap/2.4.46-sslmaster/usr/lib64' \
'CPPFLAGS= -I/usr/local/openssl64/master/include'


, where [INSTALLROOT] point to path is user $HOME tree .

You could skip all those flags related to ldap (--without-ldap)  .


For historic reasons the configuration "OpenSSL library version"  runs 
(!)  test program.
In your case config.log should show reason for failure, but it is 
expected to be due to lack of RUN-PATH flags.


You could try following configurations :

a) something like above especially variables  CPPFLAGS and LDFLAGS ( 
with extar run-path !)

b)  --with-ssl-dir= ...   --with-rpath='-Wl,,-rpath,'  , where rpath is 
for linux  or -R for unix/bsd. RUN_PATH flag is important .


Build against openssl build tree is similar to a), i.e. there is no need 
install openssl.


For the "regular" tests - particular test could be run by:
$ make -e LTESTS=try-ciphers REGRESS_TARGETS=t-exec tests


Run of  X.509 related tests - out of scope .


Regards
Roumen Petrov

**Comment by mattcaswell** [comment]
I've manged to complete the build and see the failing test. The trick was to do `export LD_LIBRARY_PATH=/usr/local/ssl/lib` before running `configure`

**Comment by mattcaswell** [comment]
Fix for this in #8604

**Comment by slontis** [comment]
@levitte see matt's comment on how to build PKIX-SSH
